### PR TITLE
Add `logfire.instrument_aws_lambda`

### PR DESCRIPTION
**Pull request by Kludex**
Suggestions on how to test this?

### CODE REVIEW

File: docs/integrations/aws-lambda.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def handler(event, context):
+    return 'Hello from Lambda'
+
+logfire.instrument_aws_lambda()  # (2)!
+```
+
+1. Remember to set the `LOGFIRE_TOKEN` environment variable on your Lambda function configuration.
+2. The `logfire.instrument_aws_lambda` function must be called after defining the handler function.
```
Comment (alexmojaki) [inline docs/integrations/aws-lambda.md]: I think this should be enforced with code, e.g. require calling `logfire.instrument_aws_lambda(handler)` or check `globals()['handler']`.

**Review by alexmojaki — COMMENTED** [review]
What have you learned from manual testing? Didn't it not work in the python runtime?

I think basic automated testing like asserting that something got patched is good enough.

**Comment by Kludex** [comment]
> What have you learned from manual testing? Didn't it not work in the python runtime?

I made it work, I had to put the instrument method after the lambda function.

**Comment by alexmojaki** [comment]
And does it do something more useful than `logfire.instrument`?

**Comment by Kludex** [comment]
> And does it do something more useful than `logfire.instrument`?

It does.

Screenshot 2024-12-10 at 13 49 18

But... I wouldn't be able to explain how useful this information is.

File: docs/integrations/aws-lambda.md (hunk 1 above)
Comment (Kludex) [reply docs/integrations/aws-lambda.md]: I tried to do it with `globals()`, but the function was not there, I guess I needed to go up the stack and get the locals up? It was easier to add the `lambda_handler`.

File: logfire/_internal/main.py [hunk 2]
```diff
@@ -1451,12 +1452,31 @@ def instrument_sqlite3(
             },
         )
 
+    def instrument_aws_lambda(self, lambda_function: Any, **kwargs: Unpack[AwsLambdaInstrumentKwargs]) -> None:
```
Comment (alexmojaki) [inline logfire/_internal/main.py]: Can this be type-hinted?

Comment (Kludex) [reply logfire/_internal/main.py]: `Callable[[Any, Any], Any]`, or maybe `Callable[[dict[str, Any], dict[str, Any]], Any]`...

Comment (Kludex) [reply logfire/_internal/main.py]: I'm not 100% sure if the one on the right covers all cases...

Comment (alexmojaki) [reply logfire/_internal/main.py]: I think the basic one on the left is fine. Also maybe a note in docs about this param.

Comment (Kludex) [reply logfire/_internal/main.py]: What do you want to mention?

Comment (alexmojaki) [reply logfire/_internal/main.py]: I guess just mention what they're supposed to pass in

Comment (Kludex) [reply logfire/_internal/main.py]: Please suggest, and apply whatever you'd like to see.

The example contains the function as first parameter...

File: logfire/_internal/main.py [hunk 3]
```diff
@@ -1451,12 +1452,33 @@ def instrument_sqlite3(
             },
         )
 
+    def instrument_aws_lambda(
+        self, lambda_function: LambdaFunction, **kwargs: Unpack[AwsLambdaInstrumentKwargs]
```
Comment (alexmojaki) [inline logfire/_internal/main.py]:
```suggestion
        self, lambda_function_handler: LambdaFunctionHandler, **kwargs: Unpack[AwsLambdaInstrumentKwargs]
```

Comment (Kludex) [reply logfire/_internal/main.py]: The examples on AWS use the `lambda_function` name for the function, I prefer to use the same name.

Comment (alexmojaki) [reply logfire/_internal/main.py]: > A function handler can be any name; however, the default name in the Lambda console is lambda_function.lambda_handler. This function handler name reflects the function name (lambda_handler) and the file where the handler code is stored (lambda_function.py).

https://docs.aws.amazon.com/lambda/latest/dg/python-handler.html

Comment (Kludex) [reply logfire/_internal/main.py]: ah yeah, `lambda_handler`, not `lambda_function`. mb

Comment (Kludex) [reply logfire/_internal/main.py]: Anything else in this PR besides this? I'll merge after fixing this otherwise.

**Pull request merged**